### ISSUE TITLE
[Bug] Protecting WebAPI does not work if one wants to use multiple bearer schemes, e.g. for AAD *and* B2C

### ISSUE THREAD

**Issue by astaykov** [AzureAD/microsoft-identity-web#468]
Using **Microsoft Identity Web 0.2.3-preview** protecting WebAPI with multiple Bearer schemes for different authorities.

Referring to [this sample for multi bearer auth for Web Apis](https://github.com/azure-ad-b2c/apps/blob/master/apps/spa-hellojs/source-code/.Net-Core-API-RBAC/MultiBearerAPI/Startup.cs), it is impossible to use Microsoft.Identity.Web package to accomplish the required setup.

Steps to reproduce:

Create the following configuration file (only relevant sections are listed here for clarity):
```
{
  "AzureAdB2C": {
    "Instance": "https://snetga.b2clogin.com/",
    "ClientId": "3422b1f4-8039-4023-9373-dbe747b40125",
    "TenantId": "126bbbd5-e975-43a7-88ac-babc84279788",
    "Domain": "snetga.onmicrosoft.com",
    "SignUpSignInPolicyId": "B2C_1A_signup_signin_google"
  },
  "AzureAd": {
    "Instance": "https://login.microsoftonline.com/",
    "TenantId": "126bbbd5-e975-43a7-88ac-babc84279788",
    "ClientId": "3422b1f4-8039-4023-9373-dbe747b40125"
  }
}
```
Then try to register both **AzureAd** and **AzureAdB2C** the following way:

```
            services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAdB2C", "B2C");
            services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAd", "v2.0");

            services.AddAuthorization(options =>
            {
                options.DefaultPolicy = new AuthorizationPolicyBuilder()
                    .RequireClaim("iss")
                    .AddAuthenticationSchemes("B2C", "v2.0")
                    .Build();
            }
            );

```

If the order is like here displayed - first registering AzureAdB2C and then registering AzureAd, you will get the following error when calling the API with valid token:

```
System.InvalidOperationException: IDX20803: Unable to obtain configuration from: 'https://login.microsoftonline.com/snetga.onmicrosoft.com/B2C_1A_signup_signin_google/v2.0/.well-known/openid-configuration'.
 ---> System.IO.IOException: IDX20807: Unable to retrieve document from: 'https://login.microsoftonline.com/snetga.onmicrosoft.com/B2C_1A_signup_signin_google/v2.0/.well-known/openid-configuration'. HttpResponseMessage: 'StatusCode: 404, ReasonPhrase: 'Not Found', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
```

If changing the order - first register the AzureAd and then AzureAdB2C, e.g.:

```
            services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAd", "v2.0");
            services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAdB2C", "B2C");
            ...

```
You will get success (200) calling the WebAPI with Azure AD B2C token, but 401 calling the API using v2.0 Token.

If you only use one of them (either AzureAd only or AzureAdB2C only) - it works.

Also, if you use the Microsoft.Identity.Web extensions to register only one of the providers (AzureAd OR AzureAdB2C), but use the Microsoft.AspNetCore.Authentication.XXX for the other provider, e.g.:

```
          services.AddAuthentication()
            .AddJwtBearer("v2.0", options =>
            {
                options.Authority = $"{Configuration["AzureAd:Instance"]}/{Configuration["AzureAd:TenantId"]}/v2.0";
                options.TokenValidationParameters = new TokenValidationParameters
                {
                    ValidateIssuer = true,
                    ValidIssuer = $"{Configuration["AzureAd:Instance"]}/{Configuration["AzureAd:TenantId"]}/v2.0",
                    ValidateAudience = true,
                    ValidAudience = Configuration["AzureAd:ClientId"],
                    ValidateLifetime = true
                };
            }); services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAd", "v2.0");
            
            services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAdB2C", "B2C");


            services.AddAuthorization(options =>
            {
                options.DefaultPolicy = new AuthorizationPolicyBuilder()
                    .RequireClaim("iss")
                    .AddAuthenticationSchemes("B2C", "v2.0")
                    .Build();
            }
            );
```

Then the API accepts both AzureAd (v2.0) **AND** AzureAdB2C tokens.

After some very short search through the code, it seems [this registration in WebApiAuthenticationBuilderExtensions](https://github.com/AzureAD/microsoft-identity-web/blob/d2ad0f5f830391a34175d48621a2c56011a45082/src/Microsoft.Identity.Web/WebApiExtensions/WebApiAuthenticationBuilderExtensions.cs#L83) might be causing the issue. Because of `builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IValidateOptions<MicrosoftIdentityOptions>, MicrosoftIdentityOptionsValidation>())`, which is trying to add a singleton of `MicrosoftIdentityOptions` 
- basically the second use of `AddMicrosoftWebApiAuthentication` will override any configuration settings that are different from the first registration. If I read correctly the [documentation about TryAddEnumerable](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.extensions.servicecollectiondescriptorextensions.tryaddenumerable?view=dotnet-plat-ext-3.1), it explains the behavior I am observing. But this is just a guess.

Expected behavior is that I am able to use `AddMicrosoftWebApiAuthentication` multiple times with different configuration settings. There are a lot of valid scenarios where one would like to have this:

 * Using Azure AD B2C for end-users *and* `client credentials` flow (which is exactly what [this sample](https://github.com/azure-ad-b2c/apps/blob/master/apps/spa-hellojs/source-code/.Net-Core-API-RBAC/MultiBearerAPI/Startup.cs) is about)
 * Having an application that authenticates users from different authorities - e.g. Commercial and China/GCC etc.

**Comment by jmprieur** [comment]
@astaykov 
By using `AddMicrosoftWebApiAuthentication`, this also sets the default scheme.

Do you try

```CSharp
services.AddAuthentication()
 .AddMicrosoftWebApi()
```

**Comment by pmaytak** [comment]
Thanks @astaykov for investigation and a write-up. 

This is also a duplicate of #429. I wrote a [reply there](https://github.com/AzureAD/microsoft-identity-web/issues/429#issuecomment-675556730).

The issue is with calling `builder.Services.Configure(configureMicrosoftIdentityOptions);` twice - with second time overwriting the first options.

`builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IValidateOptions<MicrosoftIdentityOptions>, MicrosoftIdentityOptionsValidation>());` I think is fine because it adds a singleton of a validator class, not options class.

**Comment by astaykov** [comment]
> @astaykov
> By using `AddMicrosoftWebApiAuthentication`, this also sets the default scheme.
> 
> Do you try?
> 
> ```cs
> services.AddAuthentication()
>  .AddMicrosoftWebApi(Configuration, "AzureAdB2C", "B2C")
>  .AddMicrosoftWebApi(Configuration, "AzureAd", "v2.0");
> ```
> 
> You make a good point about the options. We might need to try to use named options (with an IOptionsSnapshot) for MicrosoftIdentityOptions to support this scenario @jennyf19 and @pmaytak
> 
> @astaykov; Does your work around fully work for you? is it ok to prioritize this fix after our 0.3.0-preview?

Using the suggested approach has absolutely same (erronous) results like in my original approach.
My workaround works fully, and it is so far fine to work this way given the preview status of the package.

**Comment by pmaytak** [comment]
Thanks @astaykov for all the feedback. A fix is in PR #475 (which will be included in the next release). Also added a small [section](https://github.com/AzureAD/microsoft-identity-web/wiki/Web-APIs#using-multiple-authentication-schemes) to the wiki related to this. I'll close this issue as duplicate of #429.

**Comment by jennyf19** [comment]
Included in [0.4.0-preview release](https://github.com/AzureAD/microsoft-identity-web/releases/tag/0.4.0-preview)

**Comment by jennyf19** [comment]
Included in [0.4.0-preview release](https://github.com/AzureAD/microsoft-identity-web/releases/tag/0.4.0-preview)
